### ISSUE TITLE
Different gradient results for same custom loss function in Tensorflow 2.0

### ISSUE THREAD

**Issue by aGiant** [tensorflow/tensorflow#36694]
Tensorflow 2.0 & Python 3.7

```
def custom_loss2(output2, label2):
    loss_value = K.mean(binary_crossentropy(label2, output2)) 
    return loss_value
def EWC_loss(new_weights, old_weights, fisher_matrix, rate):
    sum_w = 0
    for v in range(len(fisher_matrix)):
        sum_w += tf.reduce_sum(tf.multiply(fisher_matrix[v], tf.square(new_weights[v] - old_weights[v]))) 
    return sum_w*rate
```

Two different trains, the same results were expected but two very different results came out. 
First one, just let rate = 0.5

```
del optimizer
del tape
optimizer = tf.keras.optimizers.SGD()
ewc_model = tf.keras.models.clone_model(model)
old_weights = model.trainable_variables.copy()
for epoch in range(num_epochs):
    with tf.GradientTape() as tape:
        out = ewc_model(features_t)
        new_weights = ewc_model.trainable_variables.copy()
        ewc_loss = EWC_loss(new_weights, old_weights, fisher_matrix, 0.5)
        loss = ewc_loss + custom_loss2(out, labels_2_t)
        grad = tape.gradient(loss, ewc_model.trainable_variables)
        optimizer.apply_gradients(grads_and_vars=zip(grad, ewc_model.trainable_variables))
    if (epoch+1)%100 == 0:
        print("epch: {}, loss: {}".format(epoch, loss.numpy()))
        print(ewc_loss.numpy(), loss.numpy())
```

In case II, just put the factor out and let rate = 1.0 but put 0.5 to ewc_loss:

```
del optimizer
del tape
optimizer = tf.keras.optimizers.SGD()
ewc_model = tf.keras.models.clone_model(model)
old_weights = model.trainable_variables.copy()
for epoch in range(num_epochs):
    with tf.GradientTape() as tape:
        out = ewc_model(features_t)
        new_weights = ewc_model.trainable_variables.copy()
        ewc_loss = 0.5*EWC_loss(new_weights, old_weights, fisher_matrix, 1.0)
        loss = ewc_loss + custom_loss2(out, labels_2_t)
        grad = tape.gradient(loss, ewc_model.trainable_variables)
        optimizer.apply_gradients(grads_and_vars=zip(grad, ewc_model.trainable_variables))
    if (epoch+1)%100 == 0:
        print("epch: {}, loss: {}".format(epoch, loss.numpy()))
        print(ewc_loss.numpy(), loss.numpy())
```

Tests were with random samples, but the same for those test codes described above. 
We evaluated the model using two test sets, loss = custom_loss2(prd, labels_2). 
The first case returned:

```
tf.Tensor(14.144677747478463, shape=(), dtype=float64)
tf.Tensor(14.254150624518838, shape=(), dtype=float64)
```

but the case II returned:

```
tf.Tensor(0.15645654679566814, shape=(), dtype=float64)
tf.Tensor(0.263113701303186, shape=(), dtype=float64)
```

Can anyone identify this issue? If the gradient was wrong, then the risk of applying tf.GradientTape() is very high using custom loss function.

**Comment by gadagashwini** [comment]
@aGiant, I tried replicating the issue, but getting different error.
Please take a look at [gist](https://colab.sandbox.google.com/gist/gadagashwini/bc099589a92797d7f66dedad888e0433/untitled392.ipynb) and provide more information to reproduce issue. Thanks!

**Comment by aGiant** [comment]
For the first run, just delete
```
del optimizer
del tape
```

Model was defined as:
```
# Input layer, one hidden layer
input_layer = Input(batch_shape=(None, 20))
dense_1 = Dense(1028)(input_layer)
output_2 = Dense(1, activation="sigmoid")(dense_1)
model = Model(inputs=input_layer, outputs= output_2)
print(model.summary())
```
Data was generated by 
```
n_sample = 1000
fix = np.array([range(n_sample),]*20).transpose()
features = np.cos(fix + np.random.rand(n_sample,20))
labels_2 = np.cos(np.array([range(n_sample),]*1).transpose() + np.random.rand(n_sample,1))
labels_2 = np.array([labels_2>=0]).astype(float)
```

**Comment by aGiant** [comment]
> @aGiant, I tried replicating the issue, but getting different error.
> Please take a look at [gist](https://colab.sandbox.google.com/gist/gadagashwini/bc099589a92797d7f66dedad888e0433/untitled392.ipynb) and provide more information to reproduce issue. Thanks!

now the error was re-produced: https://colab.research.google.com/gist/aGiant/7893ade6d21be98bebbeb91f69fc0b84/untitled392.ipynb

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.2.0-rc2 and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/e790680dd021f39d1daa0a8826c2a1ea/untitled765.ipynb). Thanks!

**Comment by jvishnuvardhan** [comment]
@aGiant This small mismatch in the results is due to cloning model. When you use clone a model, it instantiate a model and it creates new weights. [check here](https://www.tensorflow.org/api_docs/python/tf/keras/models/clone_model). Print weights for each of those two cases, you can see different weights.

> Model cloning is similar to calling a model on new inputs, except that it creates new layers (and thus new weights) instead of sharing the weights of the existing layers.

If you want exactly same results but wants to clone model, then you can use `set_weights` as follows.

`ewc_model.set_weights(model.get_weights()) # add this line to use exactly same weights `

I did exactly as mentioned above and got exactly same results from both the cases. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/25e15063ea6700582e37a51945a27f11/untitled765.ipynb).

Please close the issue if this was resolved for you. Thanks!

**Comment by aGiant** [comment]
> @aGiant This small mismatch in the results is due to cloning model. When you use clone a model, it instantiate a model and it creates new weights. [check here](https://www.tensorflow.org/api_docs/python/tf/keras/models/clone_model). Print weights for each of those two cases, you can see different weights.
> 
> > Model cloning is similar to calling a model on new inputs, except that it creates new layers (and thus new weights) instead of sharing the weights of the existing layers.
> 
> If you want exactly same results but wants to clone model, then you can use `set_weights` as follows.
> 
> `ewc_model.set_weights(model.get_weights()) # add this line to use exactly same weights `
> 
> I did exactly as mentioned above and got exactly same results from both the cases. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/25e15063ea6700582e37a51945a27f11/untitled765.ipynb).
> 
> Please close the issue if this was resolved for you. Thanks!

Many thanks for your reply!
Yes, I saw the solution there and explained the results. 

But the task is not to use the same weights but to get the almost the same convergent from any initial start points. In the end, the difference (abs error) between the two approaches should small enough, i.e. <= 1e-6. But the re-production from https://colab.research.google.com/gist/ravikyram/e790680dd021f39d1daa0a8826c2a1ea/untitled765.ipynb, error = 0.11463253099176529 >>1e-6. 
Maybe, it's because of the SGD.

**Comment by jvishnuvardhan** [comment]
@aGiant In my experience I have never seen<<1e-6 difference when complex custom_loss functions are used. There could be many reasons why it is `error = 0.11463253099176529 >>1e-6.`.  It could be due to the custom_loss or random_weights generated at the cloning inference, or SGD, or random_samples, or data size, or numerical error etc. Thanks!
Hope this helps you. Thanks!

**Comment by aGiant** [comment]
@jvishnuvardhan
Here is another test using SGD with error <= 1e-6 with many different initial start points and they all got the same error delta within the same epochs = 22.
![image](https://user-images.githubusercontent.com/25084782/78538987-24b73980-77f2-11ea-9144-9f56e16c8650.png)

With tensorflow 2.0, with SGD, same data, same setting till error<=1e-6, it took 151 ± n, with different initial start points, I got different epoch...
![image](https://user-images.githubusercontent.com/25084782/78539005-2b45b100-77f2-11ea-829c-2c555112a07f.png)

**Comment by jvishnuvardhan** [comment]
@aGiant Looks like this was resolved in recent `tf-nightly`. The losses from two approaches are same. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/7e5689d9d23203acd54f266b8e88a35d/untitled765.ipynb) and let us know whether it is resolved for you or not. 

Please verify once and close the issue if this was resolved for you. Thanks!

**Comment by jvishnuvardhan** [comment]
Closing this issue as this was resolved in `tf-nightly`. Please let us know if you face the error again, we can reopen the issue. Thanks!
